### PR TITLE
feat: upgrade network controller to v20

### PR DESCRIPTION
**Pull request by bergeron**
## **Description**

Upgrades the network controller to v20, where `providerConfig` is removed from state.

To handle this, the `getProviderConfig` selector is shimmed to return the same data as before.  Places that were accessing state directly are now sent through this selector.

A migration is added to remove `providerConfig` from state.

A helper function `mockNetworkState` is added to abstract the network schema from unit + e2e tests.  This will simplify test writing and the next network controller upgrade, where this state will change again.  

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26150?quickstart=1)

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by bergeron** [comment]
@metamaskbot update-policies

**Comment by bergeron** [comment]
@metamaskbot update-policies

**Comment by bergeron** [comment]
@metamaskbot update-policies

**Comment by bergeron** [comment]
@metamaskbot update-policies

File: test/e2e/fixture-builder.js [hunk 1]
```diff
@@ -229,15 +229,9 @@ class FixtureBuilder {
...
   withNetworkControllerOnMainnet() {
-    merge(this.fixture.data.NetworkController, {
-      providerConfig: {
-        chainId: CHAIN_IDS.MAINNET,
-        nickname: '',
-        rpcUrl: '',
-        type: NETWORK_TYPES.MAINNET,
-      },
+    return this.withNetworkController({
+      providerConfig: { id: 'mainnet' },
```
Comment (bergeron) [inline test/e2e/fixture-builder.js:233]: The occurrences of `providerConfig` must remain in this file for e2e tests to pass.

Why?  It seems like this e2e state runs through all state migrations.   If I use the new pattern with  `selectedNetworkClientId`, migration 92.2 removes it.  It used to get put back by migration 94, but that pulls it from `providerConfig`.  So keeping `providerConfig` lets us get to the right state in the end.

It feels like maybe e2e should just use the state we provide without running all migrations on it.

File: app/scripts/migrations/124.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+function transformState(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  if (
+    hasProperty(state, 'NetworkController') &&
+    isObject(state.NetworkController)
+  ) {
+    delete state.NetworkController.providerConfig;
```
Comment (bergeron) [inline app/scripts/migrations/124.ts:36]: Even with this I somehow end up with

```
    "providerConfig": {
      "ticker": "ETH"
    },
```

Not yet sure whose writing  that

File: ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts [hunk 3]
```diff
@@ -61,28 +47,6 @@ describe('useConfirmationNetworkInfo', () => {
     );
 
     expect(result.current.networkDisplayName).toBe('Custom Mainnet RPC');
-    expect(result.current.networkImageUrl).toBe('./some_image');
```
Comment (bergeron) [inline ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts:64]: `imageUrl` is a field I'm not sure about.  I don't see it anywhere in the `NetworkController`, yet there are places in extension expecting it in `providerConfig` and `networkConfigurations`.

File: ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts [hunk 4]
```diff
@@ -35,22 +29,14 @@ describe('useConfirmationNetworkInfo', () => {
...
-              rpcUrl: 'https://testrpc.com',
+            networkClientId: {
+              id: 'networkClientId',
               chainId: '0x7',
+              rpcUrl: 'https://testrpc.com',
               nickname: 'Custom Mainnet RPC',
-              type: 'rpc',
-              id: 'testNetworkConfigurationId',
-              rpcPrefs: {
-                imageUrl: './some_image',
```
Comment (bergeron) [inline ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts:52]: `imageUrl` is a field I'm not sure about.  I don't see it anywhere in the `NetworkController`, yet there are places in extension expecting it in `providerConfig` and `networkConfigurations`.

**Comment by bergeron** [comment]
@metamaskbot update-policies

File: yarn.lock [hunk 5]
```diff
@@ -4527,29 +4527,29 @@ __metadata:
...
-  checksum: 10/312c3d8777c18b30ca04a2ebf43b110b4f888fb3cfeb4f1dce6dd6b37ca5656ffa160a707fe2933bef87bcfda9e3249f21ff7dab673d3a01f660044ae8bf31e7
+  checksum: 10/9d3ba53cd41a7b725579dec3723ee92444b45c51a0ed7a283caad047c1a9d3d4c8edf2cf143c4abf8959f3ade62cc7cdcb552fe1f4bfc1ef0f6a6d6136ce958f
   languageName: node
   linkType: hard
 
 "@mantine/hooks@npm:^7.8.0":
-  version: 7.10.2
-  resolution: "@mantine/hooks@npm:7.10.2"
+  version: 7.8.0
+  resolution: "@mantine/hooks@npm:7.8.0"
```
[the remark this answers is not in the document]
Comment (bergeron) [reply yarn.lock:4549]: Hmmm not sure how to determine what caused this.  package.json only maintains or increases versions, so not sure why this went down.

File: test/data/mock-send-state.json [hunk 6]
```diff
@@ -144,17 +144,16 @@
       "0x539": true
     },
     "showTestnetMessageInDropdown": true,
-    "networkConfigurations": {},
+    "networkConfigurations": {
+      "goerli": {
+        "chainId": "0x5"
+      }
```
Comment (mikesposito) [inline test/data/mock-send-state.json:150]: nit
```suggestion
    "networkConfigurations": {
      "goerli": {
        "id": "goerli",
        "chainId": "0x5"
      }
```

File: test/e2e/tests/metrics/state-snapshots/errors-before-init-opt-in-background-state.json [hunk 7]
```diff
@@ -79,15 +79,7 @@
...
-      "providerConfig": {
-        "chainId": "0x539",
-        "nickname": "Localhost 8545",
-        "rpcPrefs": "object",
-        "rpcUrl": "string",
-        "ticker": "ETH",
-        "type": "rpc",
-        "id": "networkConfigurationId"
-      },
+      "providerConfig": "object",
```
Comment (mikesposito) [inline test/e2e/tests/metrics/state-snapshots/errors-before-init-opt-in-background-state.json:82]: Does this have to stay here for now because of migrations running in e2e?

File: test/e2e/tests/network/add-custom-network.spec.js [hunk 8]
```diff
@@ -587,7 +587,7 @@ describe('Custom network', function () {
           fixtures: new FixtureBuilder()
             .withNetworkController({
               networkConfigurations: {
-                networkConfigurationId: {
+                networkConfigurationId2: {
```
Comment (mikesposito) [inline test/e2e/tests/network/add-custom-network.spec.js:590]: Curious: why was this needed?

File: app/scripts/lib/ppom/ppom-middleware.test.ts [hunk 9]
```diff
@@ -64,7 +64,12 @@ const createMiddleware = (
   }
 
   const networkController = {
-    state: { providerConfig: { chainId: chainId || CHAIN_IDS.MAINNET } },
+    state: {
+      selectedNetworkClientId: 'networkClientId',
+      networkConfigurations: {
+        networkClientId: { chainId: chainId || CHAIN_IDS.MAINNET },
+      },
+    },
```
[the remark this answers is not in the document]
Comment (bergeron) [reply app/scripts/lib/ppom/ppom-middleware.test.ts:72]: The real state would include the ID, so yeah probably.  Even though most tests don't need it in their mocks to function.

File: privacy-snapshot.json [hunk 10]
```diff
@@ -17,6 +17,7 @@
   "connect.trezor.io",
   "customnetwork.test",
   "doesntexist.test",
+  "api.etherscan.io",
```
[the remark this answers is not in the document]
Comment (bergeron) [reply privacy-snapshot.json:20]: I haven't yet figure out what caused this.  It's a subdomain of an existing host, and it's here because the tests weren't passing without it.

File: shared/modules/selectors/smart-transactions.ts [hunk 11]
```diff
@@ -53,7 +50,7 @@ type SmartTransactionsMetaMaskState = {
     smartTransactionsState: {
       liveness: boolean;
     };
-    networkConfigurations: {
+    networkConfigurations?: {
```
[the remark this answers is not in the document]
Comment (bergeron) [reply shared/modules/selectors/smart-transactions.ts:53]: Probably, since it's being passed to selectors that reference `selectedNetworkClientId`

File: test/e2e/fixture-builder.js (hunk 1 above)
Comment (bergeron) [reply test/e2e/fixture-builder.js:233]: Good eye, I hadn't noticed `FIXTURE_STATE_METADATA_VERSION` is the reason for this

File: test/e2e/tests/metrics/state-snapshots/errors-before-init-opt-in-background-state.json (hunk 7 above)
Comment (bergeron) [reply test/e2e/tests/metrics/state-snapshots/errors-before-init-opt-in-background-state.json:82]: I think so, yes.  I think same reason discussed in https://github.com/MetaMask/metamask-extension/pull/26150#discussion_r1695431277

File: test/e2e/tests/network/add-custom-network.spec.js (hunk 8 above)
Comment (bergeron) [reply test/e2e/tests/network/add-custom-network.spec.js:590]: Without this, the test fails because arbitrum becomes the current network which cannot be deleted. I can give the ID a prettier name.

**Comment by bergeron** [comment]
@metamaskbot update-policies

**Comment by bergeron** [comment]
@metamaskbot update-policies

File: test/e2e/fixture-builder.js [hunk 12]
```diff
@@ -40,32 +40,15 @@ function onboardingFixture() {
...
-            ticker: 'ETH',
-            networkConfigurationId: 'networkConfigurationId',
-            type: 'rpc',
-          },
-        },
+          rpcUrl: 'http://localhost:8545',
+          ticker: 'ETH',
+          blockExplorerUrl: undefined,
+        }),
+        providerConfig: { id: 'networkConfigurationId' },
```
Comment (bergeron) [inline test/e2e/fixture-builder.js:51]: See https://github.com/MetaMask/metamask-extension/pull/26150#discussion_r1695431277 for why these `providerConfig`s remain in the e2e mocks

File: .yarn/patches/@metamask-network-controller-npm-20.2.0-98b1a5ae59.patch [hunk 13]
```diff
@@ -0,0 +1,33 @@
+diff --git a/PATCH.txt b/PATCH.txt
+new file mode 100644
+index 0000000000000000000000000000000000000000..78b9156dc2b0bf7c33dadf325cb3ec0bfae71ccb
+--- /dev/null
++++ b/PATCH.txt
+@@ -0,0 +1,3 @@
++We remove `lookupNetwork` from `initializeProvider` in the network controller to prevent network requests before user onboarding is completed.
++The network lookup is done after onboarding is completed, and when the extension reloads if onboarding has been completed.
++This patch is part of a temporary fix that will be reverted soon to make way for a more permanent solution. https://github.com/MetaMask/metamask-extension/pull/23005
```
Comment (bergeron) [inline .yarn/patches/@metamask-network-controller-npm-20.2.0-98b1a5ae59.patch:9]: I don't know the context of this patch, but it was here before so I'm maintaining it.

File: privacy-snapshot.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (mikesposito) [reply privacy-snapshot.json:21]: Do you recall which test case was not passing? needing to add this is usually a sign that we are hitting some production API during e2e, which is usually due to running tests outside of a purely e2e env.

I just tried removing the line and running `yarn user-actions-benchmark:chrome` and it worked without issues, but maybe there's some other case which fails

File: ui/components/multichain/asset-picker-amount/asset-picker/asset-picker.test.tsx
Comment (mikesposito) [inline ui/components/multichain/asset-picker-amount/asset-picker/asset-picker.test.tsx:1]: Was mainnet already used in this test case, or is it something we are starting to do in this PR?

File: ui/pages/confirmations/components/simulation-details/fiat-display.test.tsx [hunk 14]
```diff
@@ -5,17 +5,10 @@ import { merge } from 'lodash';
...
 import { FIAT_UNAVAILABLE } from './types';
 
-const mockStateWithTestnet = merge({}, mockState, {
-  metamask: {
-    providerConfig: {
-      chainId: CHAIN_IDS.SEPOLIA,
-    },
-  },
-});
+const mockStateWithTestnet = merge({}, mockState, {});
```
Comment (mikesposito) [inline ui/pages/confirmations/components/simulation-details/fiat-display.test.tsx:11]: nit: `mockState` could be used directly here

File: ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts (hunk 4 above)
Comment (mikesposito) [reply ui/pages/confirmations/hooks/useConfirmationNetworkInfo.test.ts:52]: Previous versions of NetworkController also didn't include this field in the state

File: ui/pages/settings/networks-tab/networks-form/use-safe-chains.ts [hunk 15]
```diff
@@ -49,8 +49,12 @@ export const rpcIdentifierUtility = (
...
   for (const chain of safeChains) {
     for (const rpc of chain.rpc) {
-      if (host === new URL(rpc).host) {
-        return host;
+      try {
+        if (host === new URL(rpc).host) {
+          return host;
+        }
+      } catch {
+        continue;
```
Comment (mikesposito) [inline ui/pages/settings/networks-tab/networks-form/use-safe-chains.ts:57]: curious: why was this `try` needed?

File: ui/pages/settings/networks-tab/networks-tab-content/networks-tab-content.test.js [hunk 16]
```diff
@@ -3,22 +3,14 @@ import configureMockStore from 'redux-mock-store';
...
-      rpcUrl: 'http://localhost:8545',
-      ticker: 'ETH',
-      type: 'localhost',
-    },
-    orderedNetworkList: {
-      chainId: '0x539',
-      rpcUrl: 'http://localhost:8545',
-    },
+    ...mockNetworkState({ chainId: CHAIN_IDS.LOCALHOST }),
+    orderedNetworkList: {},
```
Comment (mikesposito) [inline ui/pages/settings/networks-tab/networks-tab-content/networks-tab-content.test.js:13]: Should `orderedNetworkList` still include entries for the used chain?

File: ui/pages/settings/networks-tab/networks-tab.test.js [hunk 17]
```diff
@@ -1,23 +1,14 @@
...
-      ticker: 'ETH',
-      type: 'localhost',
-    },
-    networkConfigurations: {},
-    orderedNetworkList: {
-      chainId: '0x539',
-      rpcUrl: 'http://localhost:8545',
-    },
+    ...mockNetworkState({ chainId: CHAIN_IDS.LOCALHOST }),
+    orderedNetworkList: {},
```
Comment (mikesposito) [inline ui/pages/settings/networks-tab/networks-tab.test.js:11]: Same as the above comment, do you think `orderedNetworkList` should still be populated?

**Review by mikesposito — COMMENTED** [review]
Amazing work! I left some comments but mostly questions and nits - my main open questions are related to `api.etherscan.io` being added to `privacy-snapshot.json` and the doubt we are starting to use `mainnet` or production apis during e2e or unit tests.

Though, this PR looks in great shape!

File: test/stub/networks.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        ? network.rpcUrl
+        : `https://localhost/rpc/${network.chainId}`,
+    nickname:
+      'nickname' in network
+        ? network.nickname
+        : (NETWORK_TO_NAME_MAP as Record<Hex, string>)[network.chainId],
+    ticker:
+      'ticker' in network
+        ? network.ticker
+        : (CHAIN_ID_TO_CURRENCY_SYMBOL_MAP as Record<Hex, string>)[
```
[the remark this answers is not in the document]
Comment (bergeron) [reply test/stub/networks.ts:39]: Yes 
Screenshot 2024-08-20 at 8 21 23 AM

File: test/e2e/tests/tokens/token-list.spec.ts [hunk 19]
```diff
@@ -97,7 +97,7 @@ describe('Token List', function () {
     );
   });
 
-  it.only('shows percentage increase for an ERC20 token with prices available', async function () {
+  it('shows percentage increase for an ERC20 token with prices available', async function () {
```
[the remark this answers is not in the document]
Comment (bergeron) [reply test/e2e/tests/tokens/token-list.spec.ts:100]: I assumed the `.only` was unintentional.  The rest of the tests pass, so I figured I'd turn them back on.

File: test/e2e/tests/transaction/send-eth.spec.js [hunk 20]
```diff
@@ -303,7 +303,7 @@ describe('Send ETH', function () {
             });
 
             // the transaction has the expected gas price
-            driver.clickElement(
+            await driver.clickElement(
```
[the remark this answers is not in the document]
Comment (bergeron) [reply test/e2e/tests/transaction/send-eth.spec.js:306]: I don't think the fix is specific this upgrade.  I was just debugging failures in various test suites, and while I was there I noticed missing awaits

File: test/stub/networks.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    id: network.id ?? uuidv4(),
+    chainId: network.chainId,
+    rpcUrl:
+      'rpcUrl' in network
+        ? network.rpcUrl
+        : `https://localhost/rpc/${network.chainId}`,
+    nickname:
+      'nickname' in network
+        ? network.nickname
+        : (NETWORK_TO_NAME_MAP as Record<Hex, string>)[network.chainId],
```
[the remark this answers is not in the document]
Comment (bergeron) [reply test/stub/networks.ts:35]: Yes similar to https://github.com/MetaMask/metamask-extension/pull/26150#discussion_r1723509595

File: test/e2e/tests/network/custom-rpc-history.spec.js [hunk 22]
```diff
@@ -186,24 +188,21 @@ describe('Custom RPC history', function () {
...
                 rpcUrl: 'http://127.0.0.1:8545/2',
                 chainId: '0x539',
                 ticker: 'ETH',
                 nickname: 'http://127.0.0.1:8545/2',
-                rpcPrefs: {},
-                type: 'rpc',
               },
-            },
+            ),
+            selectedNetworkClientId: 'mainnet',
```
[the remark this answers is not in the document]
Comment (bergeron) [reply test/e2e/tests/network/custom-rpc-history.spec.js]: Part of the behavior of `mockNetworkState` (which I should JSDoc), is that it sets the first network in the array to the currently selected network.  But to match the existing behavior of this test, we just want to add additional networks without affecting which one is current.   I've made that more explicit in https://github.com/MetaMask/metamask-extension/pull/26150/commits/e50b5bb69b41936d2f91f209330149b5b16217f3

File: ui/pages/settings/networks-tab/networks-form/use-safe-chains.ts (hunk 15 above)
Comment (bergeron) [reply ui/pages/settings/networks-tab/networks-form/use-safe-chains.ts:57]: One of the rpc URLs was not a valid URL (they're sourced from `https://chainid.network/chains.json`).  It was something like `https://foo.bar, wss://foo.bar`.  This threw an error and prevented the rest of the loop from running.

File: ui/components/app/assets/nfts/nfts-tab/nfts-tab.test.js [hunk 23]
```diff
@@ -183,7 +181,8 @@ const render = ({
         [CHAIN_IDS.MAINNET]: {},
         [CHAIN_IDS.GOERLI]: {},
       },
-      providerConfig: { chainId, type: NETWORK_TYPES.MAINNET },
+      ...mockNetworkState({ chainId }),
+      currencyRates: {},
```
[the remark this answers is not in the document]
Comment (bergeron) [reply ui/components/app/assets/nfts/nfts-tab/nfts-tab.test.js:185]: Because the mocks include `ticker: ETH` now.  It could also be fixed by explicitly doing `ticker: undefined`

File: ui/components/multichain/asset-picker-amount/asset-picker/asset-picker.test.tsx
Comment (bergeron) [reply ui/components/multichain/asset-picker-amount/asset-picker/asset-picker.test.tsx:1]: mainnnet was already the default chain, but I moved 2 tests to an unknown chain id to easier replicate their snapshots

File: ui/pages/confirmations/components/simulation-details/fiat-display.test.tsx (hunk 14 above)
Comment (bergeron) [reply ui/pages/confirmations/components/simulation-details/fiat-display.test.tsx]: put this back to sepolia to match previous mock: https://github.com/MetaMask/metamask-extension/pull/26150/commits/12ca1d3867cf8f370d147693d867f24b87bfc86e

File: ui/pages/settings/networks-tab/networks-tab.test.js (hunk 17 above)
Comment (bergeron) [reply ui/pages/settings/networks-tab/networks-tab.test.js]: ah yeah, i put it back to the way it was before in https://github.com/MetaMask/metamask-extension/pull/26150/commits/059ee02788cac2848847bb12f0dbb5094c578563

File: ui/pages/settings/networks-tab/networks-tab-content/networks-tab-content.test.js (hunk 16 above)
Comment (bergeron) [reply ui/pages/settings/networks-tab/networks-tab-content/networks-tab-content.test.js]: ah yeah, i put it back to the way it was before in https://github.com/MetaMask/metamask-extension/pull/26150/commits/059ee02788cac2848847bb12f0dbb5094c578563

File: privacy-snapshot.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (bergeron) [reply privacy-snapshot.json]: You're right, I removed it in https://github.com/MetaMask/metamask-extension/pull/26150/commits/9ff56ed1224abd0aa2fef5ce9926fb036042fbf3 and the tests still pass

File: yarn.lock (hunk 5 above)
[the remark this answers is not in the document]
Comment (bergeron) [reply yarn.lock:4549]: if yarn.lock is autogenerated, shouldn't I let it have the version it wants?  I'm not sure how to determine why it changed or whether its safe to override.

File: app/scripts/controllers/mmi-controller.test.ts [hunk 24]
```diff
@@ -98,13 +98,6 @@ describe('MMIController', function () {
           'NetworkController:infuraIsUnblocked',
         ],
       }),
-      state: {
-        providerConfig: {
-          type: NETWORK_TYPES.SEPOLIA,
-          chainId: CHAIN_IDS.SEPOLIA,
-          ticker: TEST_NETWORK_TICKER_MAP[NETWORK_TYPES.SEPOLIA],
-        },
-      },
```
Comment (salimtb) [inline app/scripts/controllers/mmi-controller.test.ts:107]: why this hasn't been replaced with the new state and has instead simply been deleted?

File: app/scripts/migrations/126.ts [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+function transformState(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  if (
+    hasProperty(state, 'NetworkController') &&
+    isObject(state.NetworkController)
+  ) {
+    delete state.NetworkController.providerConfig;
```
Comment (salimtb) [inline app/scripts/migrations/126.ts:36]: Currently, the function attempts to delete providerConfig without checking if it exists. To make the code more robust and avoid potential runtime errors, please add a check to verify that providerConfig exists on state.NetworkController before attempting to delete it. This can be done using `hasProperty` or by directly checking `state.NetworkController.providerConfig`.

File: app/scripts/migrations/126.test.ts [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+
+    const transformedState = await migrate({
+      meta: { version: oldVersion },
+      data: oldState,
+    });
+
+    expect(transformedState.data).toStrictEqual(oldState);
+  });
+
+  it('Removes providerConfig from the network controller state', async () => {
```
Comment (salimtb) [inline app/scripts/migrations/126.test.ts:32]: To ensure that the new logic correctly handles the presence and absence of the providerConfig property, please add unit tests that cover both scenarios: one where providerConfig exists and is successfully deleted, and another where it doesn't exist and return the existing state ( see the comment after )

**Comment by salimtb** [comment]
I've completed QA on the following scenarios, and all tests have passed:
1. Switching networks in the dapp context
2. Adding a network from a dapp
3. Adding a popular network
4. Adding a custom network using the form
5. Processing a transaction using the new state
Everything is functioning as expected.

File: test/stub/networks.ts [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+  NetworkStatus,
+} from '@metamask/network-controller';
+import { v4 as uuidv4 } from 'uuid';
+import { Hex } from '@metamask/utils';
+import {
+  NETWORK_TO_NAME_MAP,
+  CHAIN_ID_TO_CURRENCY_SYMBOL_MAP,
+} from '../../shared/constants/network';
+
+export const mockNetworkState = (
```
Comment (salimtb) [inline test/stub/networks.ts:13]: can we add unit test for this function

File: ui/ducks/metamask/metamask.js [hunk 28]
```diff
@@ -278,10 +277,17 @@ export const getAlertEnabledness = (state) => state.metamask.alertEnabledness;
...
-  return state.metamask.providerConfig;
+  const networkClientId = getSelectedNetworkClientId(state);
+  const builtInNetwork = BUILT_IN_INFURA_NETWORKS[networkClientId];
+  return builtInNetwork
+    ? {
+        ...builtInNetwork,
+        type: networkClientId,
+        rpcPrefs: { blockExplorerUrl: builtInNetwork.blockExplorerUrl },
+      }
+    : { ...getNetworkConfigurations(state)[networkClientId], type: 'rpc' };
```
Comment (salimtb) [inline ui/ducks/metamask/metamask.js:290]: you can use `NETWORK_TYPES.RPC` instead if `'rpc'` here

**Comment by salimtb** [comment]
I've completed QA on the following scenarios, and all tests have passed:

- Switching networks in the dapp context
- Adding a network from a dapp
- Adding a popular network
- Adding a custom network using the form
- Processing a transaction using the new state
- migrate from old state to the new state
- Delete network

Everything is functioning as expected ✅

File: app/scripts/migrations/126.ts (hunk 25 above)
Comment (bergeron) [reply app/scripts/migrations/126.ts:36]: That isn't necessary.  `delete` no-ops if the field does not exist.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete

`If the property which you are trying to delete does not exist, delete will not have any effect and will return true.`

File: app/scripts/migrations/126.test.ts (hunk 26 above)
Comment (bergeron) [reply app/scripts/migrations/126.test.ts:32]: There's already tests for both cases.  providerConfig being in state and not in state

`Does nothing if providerConfig is not in the network controller state`

**Comment by bergeron** [comment]
@metamaskbot update-policies

File: app/scripts/controllers/mmi-controller.test.ts (hunk 24 above)
Comment (bergeron) [reply app/scripts/controllers/mmi-controller.test.ts:107]: I think just because it passed without it.  But I put back an explicit sepolia mock in https://github.com/MetaMask/metamask-extension/pull/26150/commits/ef2eff3cc14deef6b28a9e92bab089abde0a163e

File: ui/ducks/metamask/metamask.js (hunk 28 above)
Comment (bergeron) [reply ui/ducks/metamask/metamask.js]: https://github.com/MetaMask/metamask-extension/pull/26150/commits/8d6d7be44fad16fd637792ffd94956a85ada6c54

**Review by salimtb — DISMISSED** [review]

File: test/e2e/fixture-builder.js (hunk 12 above)
Comment (bergeron) [reply test/e2e/fixture-builder.js:51]: In our v21+ upgrade branch, I bumped `FIXTURE_STATE_METADATA_VERSION` and it wasn't too difficult.  So we may be able to do it here.

It may have been responsible for jumbling the state expected by sentry tests.

**Review by sleepytanya — DISMISSED** [review]

**Review by zone-live — DISMISSED** [review]

**Review by shane-t — DISMISSED** [review]

**Review by hmalik88 — DISMISSED** [review]

**Review by mikesposito — DISMISSED** [review]

**Comment by bergeron** [comment]
https://github.com/MetaMask/metamask-extension/pull/26147 from develop breaks when merged in since it usses more provider config.  Fixing which will reset votes

**Review by hmalik88 — DISMISSED** [review]

File: yarn.lock (hunk 5 above)
[the remark this answers is not in the document]
Comment (bergeron) [reply yarn.lock]: Thanks for the help and explanation.  I didn't realize it wasn't just directly computed from package.json.  I've run those steps and yarn.lock makes more sense now, with versions increasing.  https://github.com/MetaMask/metamask-extension/pull/26150/commits/523d3d4af657e0d74fdc659a36dbf511eb81b095

**Comment by bergeron** [comment]
Thanks @legobeat, I've used those steps to regenerate yarn.lock in https://github.com/MetaMask/metamask-extension/pull/26150#discussion_r1728105579

**Comment by bergeron** [comment]
@metamaskbot update-policies

**Review by salimtb — APPROVED** [review]

**Review by zone-live — APPROVED** [review]

File: .yarn/patches/@metamask-smart-transactions-controller-npm-11.0.0-83133fb677.patch [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+index cd9bd16a3b0ddbb49de6bc7bd8fceefb8b56ad5e..066e2e132e55f2dcc9a8add4197b8697c2ff4d90 100644
+--- a/dist/SmartTransactionsController.js
++++ b/dist/SmartTransactionsController.js
+@@ -77,8 +77,8 @@ class SmartTransactionsController extends polling_controller_1.StaticIntervalPol
+         this.getMetaMetricsProps = getMetaMetricsProps;
+         this.initializeSmartTransactionsForChainId();
+         __classPrivateFieldGet(this, _SmartTransactionsController_instances, "m", _SmartTransactionsController_ensureUniqueSmartTransactions).call(this);
+-        onNetworkStateChange(({ providerConfig: newProvider }) => {
+-            const { chainId } = newProvider;
++        onNetworkStateChange((state) => {
```
[the remark this answers is not in the document]
Comment (bergeron) [reply .yarn/patches/@metamask-smart-transactions-controller-npm-11.0.0-83133fb677.patch:11]: yes that can be bumped now.  I'll do that as quick follow up so i can get the bulk of this merged

**Review by jpuri — APPROVED** [review]
Changes in files owned by confirmations team look good 👍

**Review by infiniteflower — APPROVED** [review]

File: .yarn/patches/@metamask-smart-transactions-controller-npm-11.0.0-83133fb677.patch (hunk 29 above)
[the remark this answers is not in the document]
Comment (bergeron) [reply .yarn/patches/@metamask-smart-transactions-controller-npm-11.0.0-83133fb677.patch:11]: bumped here https://github.com/MetaMask/metamask-extension/pull/26644
